docs/CONTRACTS_AND_TESTING.md

```diff
@@ -0,0 +1,899 @@
+  # Note: ~= is approximate equality operator for floating-point comparisons
+=
+  a / b
+
+# Calling with invalid input
+let x = divide 10.0 0.0  # Error: precondition violated: "divisor cannot be zero"
+```
+
+### Multiple Conditions
+
+```cadenza
+fn binary_search (arr: List Integer) (target: Integer) -> Integer
+  @requires { 
+    is_sorted arr "array must be sorted",
+    length arr > 0 "array cannot be empty"
+  }
+  @ensures {
+    result >= -1 "result is valid index or -1",
+    result < length arr "result is within bounds",
+    result >= 0 ==> get arr result == target "if found, element matches target"
+  }
+=
+  # implementation
+  ...
+```
+
+### Invariants
+
+Type invariants ensure that values always satisfy certain properties:
+
+```cadenza
+struct BankAccount {
+  balance = Float @ { x >= 0.0 },
+  owner = String,
+}
+  @invariant { balance >= 0.0 "balance cannot be negative" }
+
+# Methods must preserve invariants
+fn withdraw (account: BankAccount) (amount: Float) -> BankAccount
+  @requires { amount > 0.0 && amount <= account.balance }
+  @ensures { result.balance == account.balance - amount }
+=
+  { ...account, balance = account.balance - amount }
+
+# This would violate the invariant and be rejected
+fn break_account (account: BankAccount) -> BankAccount =
+  { ...account, balance = -100.0 }  # Error: violates balance >= 0.0 invariant
+```
+
+### Loop Invariants (Future)
+
+For formal verification of loops:
+
+```cadenza
+fn sum_array (arr: Array n Integer) -> Integer
+  @ensures { result == sum_of_elements arr }
+=
+  let total = 0
+  let i = 0
+  while i < n
+    @invariant { 
+      0 <= i && i <= n,
+      total == sum (take i arr)
+    }
+  do
+    total = total + arr[i]
+    i = i + 1
+  total
+```
+
+### Attribute Syntax
+
+Contract attributes use `@name { ... }` syntax:
+
+- `@requires { predicate }` - Precondition (checked on entry)
+- `@ensures { predicate }` - Postcondition (checked on exit, can reference `result`)
+- `@invariant { predicate }` - Type invariant (checked on construction and mutation)
+- `@modifies { variable_list }` - Documents side effects (future)
+
+---
+
+## Built-in Testing Framework
+
+### Motivation
+
+Testing should be integrated into the language, not a separate tool. Inspired by Rust's `#[test]` attribute and built-in test runner.
+
+### Test Syntax
+
+Mark functions as tests using `@test` attribute:
+
+```cadenza
+@test
+fn test_addition =
+  assert 1 + 1 == 2
+  assert 2 + 2 == 4
+
+@test "edge case: zero"
+fn test_zero =
+  assert 0 + 0 == 0
+  assert 5 + 0 == 5
+
+@test "negative numbers"
+fn test_negatives =
+  assert -1 + 1 == 0
+  assert -5 + -3 == -8
+```
+
+### Test Organization
+
+Tests are discovered automatically in the codebase. Tests can be placed in the same file as the code being tested, or in separate test files using the `.test.cdz` extension convention.
+
+```cadenza
+# In module: math.cdz
+
+fn add (a: Integer) (b: Integer) = a + b
+
+# Tests in the same file
+@test
```

Comment:
One thing I would like to have for tests is user tags so you can categorize test types and potentially filter exclude those as part of a run